### PR TITLE
Move HTTP Permissions and Roles policies from build-time to runtime

### LINKED ISSUE #12594: Quarkus Security issue in the Native Executable
Quarkus version: 1.8.1

I have two quarkus profile configurations for our security. One for the `prod` and another for the `insecure` profile:

```
quarkus.http.auth.permission.auth.paths=/*

quarkus.http.auth.permission.auth.policy=authenticated
%insecure.quarkus.http.auth.permission.auth.policy=permit
```

So what happens:
1. Set env var `QUARKUS_PROFILE="insecure"` 
2. Build the native executable 
3. Set env var `QUARKUS_PROFILE="prod"`
4. Run the native executable

Result: 
My endpoints are not secured

Expected result:
Security setting much change depending on the runtime profile 

So my security setting for the "prod" profile is not applied and all the endpoints are not secured. It seems that when building the native executable the settings from the "build profile" (the profile that is set during build time "insecure") are recorded and there is not way to change them during runtime with a different profile. 
If I'm not building a native executable and I'm running Quarkus in JVM mode, everything works as expected and when I change the runtime profile, security is applied. 
We need to have different security settings (for each quarkus profile) for our different envirouments. 
How can I do that?

### PR DESCRIPTION
**Pull request by michalvavrik**
fixes: #12594

### CODE REVIEW

File: docs/src/main/asciidoc/security-authorize-web-endpoints-reference.adoc
Comment (sberyozkin) [inline docs/src/main/asciidoc/security-authorize-web-endpoints-reference.adoc:50]: Today users can already register `HttpSecurityPoliicy` - in which case it will apply to all methods. 
I'm not sure a new interface is needed, can `HttpSecurityPoliicy` have a default `name()` returning null instead ?

By the way, it is not really related to the actual problem this PR (nicely) resolves. IMHO it just complicates things, since making it runtime only is a major change of its own, and while having a named custom policy is very interesting, IMHO it should be done independently of this PR - this way we'll have 2 real noteworthy features as well

Comment (michalvavrik) [reply docs/src/main/asciidoc/security-authorize-web-endpoints-reference.adoc:50]: > Today users can already register HttpSecurityPoliicy - in which case it will apply to all methods.
I'm not sure a new interface is needed, can HttpSecurityPoliicy have a default name() returning null instead ?

I originally had exactly that, but the `PathMatchingHttpSecurityPolicy` is a global `HttpSecurityPolicy`, so injecting `HttpSecurityPolicy` and checking for `name` will lead to circular dependency. I suppose I can analyze `BeanInfo` or something like that for bean class, but it's far from elegant.

> while having a named custom policy is very interesting, IMHO it should be done independently of this PR - this way we'll have 2 real noteworthy features as well

I think you misunderstood this. It is not a new thing, previously you could do that via build item, now since the config is runtime, it doesn't make sense, so I allowed it as a CDI bean.

Comment (michalvavrik) [reply docs/src/main/asciidoc/security-authorize-web-endpoints-reference.adoc:50]: Just to be clear - I don't dismiss your comment, I agree in general, but I also need to keep backward compatibility, so it needs to be done in this PR or I'll have to introduce some hack just for backwards compatibility. I'll wait to hear your opinion before I make any changes.

Comment (michalvavrik) [reply docs/src/main/asciidoc/security-authorize-web-endpoints-reference.adoc:50]: Well, considering it's a `Singleton`, I'll only have to do it once, so I can check for class name. Alright, I'll get rid of `NamedHttpSecurityPolicy`.

Comment (michalvavrik) [reply docs/src/main/asciidoc/security-authorize-web-endpoints-reference.adoc]: Done.

Comment (sberyozkin) [reply docs/src/main/asciidoc/security-authorize-web-endpoints-reference.adoc]: Thanks @michalvavrik, sorry if I missed some of your ideas behind it, I just thought if someone already implemented `HttpSecurityPolicy` and now finds out about this new enhancement (which I absolutely support, very nice idea :+1: ) and want to apply this policy to specific paths only, they would have to change interface as opposed to introducing a name property which appears to be simpler. I think we have something similar for one of the other HTTP related security classes

**Review by sberyozkin — COMMENTED** [review]
I've looked through and it looks perfect, but I'll need a few more iterations to make sure I follow it. Thanks Michal, nicely researched and done

**Comment by sberyozkin** [comment]
@stuartwdouglas This is a very important PR done by Michal, have a look if you get a chance, LGTM but I haven't looked very deeply yet

**Comment by sberyozkin** [comment]
@michalvavrik So what do you think about making the named policy concept realized in a follow up PR ? I recommend but not insist, IMHO it would be better - it is an independent enhancement of its own. I'll let you decide

**Comment by michalvavrik** [comment]
> @michalvavrik So what do you think about making the named policy concept realized in a follow up PR ? I recommend but not insist, IMHO it would be better - it is an independent enhancement of its own. I'll let you decide

Problem is that we don't know whether someone else is not using `io.quarkus.vertx.http.deployment.HttpSecurityPolicyBuildItem`. This PR transform the build item to the named policy and I'd like to initialize HTTP permissions inside `AbstractPathMatchingHttpSecurityPolicy` constructor because of what I am preparing for #14047. So if you want to avoid named policy, I can transform the build item for example to a package-private bean class that users won't be able to create them via CDI...

>  I'll let you decide

If you let me decide, then I think because of the `io.quarkus.vertx.http.deployment.HttpSecurityPolicyBuildItem` it is easiest to keep it in this PR. But if you or Stuart find this PR hard to review, I'll drop it. Let me know if you change your mind, it's not a big deal.

**Comment by sberyozkin** [comment]
@michalvavrik Thanks, I'd find it easier to isolate the changes related to the configuration scope only if the named policy enhancement was done in another PR, it is hard to be certain otherwise if a given change is related to the move to the runtime scope or the named policy update

**Review by sberyozkin — APPROVED** [review]

**Comment by sberyozkin** [comment]
I don't see anything concerning now, fine PR, important enhancement, lets have it settled a bit in 3.6.0 CR1

**Pull request merged**
